src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java

```diff
@@ -1228,17 +1233,31 @@ else if (!referencePath.isDirectory())
                     perModuleArgs.add(sModulePath);
                     StandardCredentials finalCred = cred;
                     URIish finalUrIish = urIish;
-                    executorService.submit(() -> {
+
+                    launchCommands.add(() -> {
                         try {
-                            launchCommandWithCredentials(perModuleArgs, workspace, finalCred, finalUrIish, timeout);
+                            return launchCommandWithCredentials(perModuleArgs, workspace, finalCred, finalUrIish, timeout);
                         } catch (InterruptedException e) {
                             throw new GitException("Interrupted while updating submodule for " + sModuleName);
                         }
                     });
                 }
 
+                List<Future<String>> futures = executorService.invokeAll(launchCommands);
                 executorService.shutdown();
```

Comment:
About the complete shutdown procedure I am not sure yet. It requires some more thought and testing.
Please have a look at ExecutorService [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html) under "Usage Examples" -> the two phase shutdown. The method `shutdownNow()` should actually be called to really try to stop running processes. The method `shutdown()` just prevents new ones to start running.

Reply:
I'm pretty sure that we want `shutdown`, and not `shutdownNow`. From the documentation:

> The shutdown() method will allow previously submitted tasks to execute before terminating, while the shutdownNow() method prevents waiting tasks from starting and attempts to stop currently executing tasks.

Our test case is quite simple, so let's imagine a larger example. Let's say the repository has 30 submodules, and the thread pool has 5 threads. So the first 5 get scheduled right away when calling `invokeAll`. If we call `shutdownNow`, these first 5 commands run to completion, but not the next 25, since they are in the waiting state. These commands will be cancelled, which is not what we want. Likewise, we don't want to stop any running commands either. Basically, we just want to block until everything finishes executing.

Therefore, we do indeed want `shutdown`, since the idea is we want all waiting tasks to finish executing. FWIW, I can verify that this behavior works in the above example, since I built HPI plugins for this plugin and `git-plugin` and tested them on a Jenkins instance with one of the big repositories at my company, which has >30 submodules. Updating submodules went much faster than serially, and all submodules were correctly updated in the checkout phase of the job.

Reply:
To elaborate a bit more:

Method `shutdown()` should be called right after submitting all commands, to prevent new submissions. It is similar to making the `ExecutorService` immutable, as it rejects any new tasks afterwards. As the `ExecutorService` has just method scope here and "we know what we are doing", calling this method would not be required.

Method `shutdownNow()` should be called before leaving the `execute()` method, to try to stop any running task. This should be done after performing the `Future::get` calls, otherwise running tasks may be stopped. `Future::get` is blocking, so after retrieving the results all tasks should anyway have been finished, as a timeout is used inside. So calling `shutdownNow()` is mainly for handling error cases.

These points should be assured and tested:
* `Exceptions` are mapped properly from the `Future::get` to maintain existing behaviour
* `ExecutorService` is shut down with all its running threads in any case to prevent thread leakage
* `ExectorService` and its threads have to be used in a way that doesn't hinder Jenkins to shutdown

Reply:
@nre-ableton BTW, are you aware of the shallow submodule pull request #348 that got merged some days ago. This should reduce execution time for submodules even more than parallel updates.

Reply:
@darxriggs I assume that you mean #344, not #348, right? Anyways, yes, I'm aware of it, and it's a great improvement to the plugin. For the repositories that I work with, parallelization is still more important due to the large number of submodules. This is mostly due to the fact that git performance on Windows is quite poor (we must build on Linux/Mac/Win in parallel). Shallow cloning is a bit faster on Windows, but not by much.

Reply:
Right, #344 is what I meant.

Reply:
@darxriggs Ok, I didn't realize that `invokeAll` would block until all threads completed (my Java is a bit rusty, apologies). Anyways I agree that in this case we should use `shutdownNow`.